clients/src/main/java/org/apache/kafka/common/record/CompressionType.java

```diff
@@ -179,4 +185,17 @@ private static MethodHandle findConstructor(String className, MethodType methodT
         }
     }
 
+    static class RecyclingBufferSupplier implements KafkaLZ4BlockInputStream.BufferSupplier {
+
+        ByteBuffer theBuffer;
+
+        @Override
+        public ByteBuffer get(int size) {
+            if (theBuffer == null || theBuffer.capacity() < size) {
+                theBuffer = ByteBuffer.allocate(size);
+            }
+            theBuffer.limit(size);
+            return theBuffer;
```

Comment:
@hachikuji technically we don't have to, KafkaLZ4BlockInputStream resets position and limit every time anyway when preparing the buffer for consumption.

@ijuma agree that would be a concern if someone were to consume from two separate consumers in the same thread. Happy to replace the buffer supplier implementation with the one you pointed to.